sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -191,25 +191,34 @@ object ExtractFiltersAndInnerJoins extends PredicateHelper {
 
 /**
  * A pattern that collects all adjacent unions and returns their children as a Seq.
+ * If the top union is wrapped in a [[Distinct]], then the [[Distinct]] in the adjacent unions, if
+ * any, will be eliminated.
  */
 object Unions {
-  def unapply(plan: LogicalPlan): Option[Seq[LogicalPlan]] = plan match {
-    case u: Union => Some(collectUnionChildren(mutable.Stack(u), Seq.empty[LogicalPlan]))
+  def unapply(plan: LogicalPlan): Option[(Seq[LogicalPlan], Boolean)] = plan match {
+    case u: Union =>
+      Some(collectUnionChildren(mutable.Stack(u), Seq.empty[LogicalPlan], false), false)
+    case Distinct(u: Union) =>
+      Some(collectUnionChildren(mutable.Stack(u), Seq.empty[LogicalPlan], true), true)
     case _ => None
   }
 
   // Doing a depth-first tree traversal to combine all the union children.
   @tailrec
   private def collectUnionChildren(
       plans: mutable.Stack[LogicalPlan],
-      children: Seq[LogicalPlan]): Seq[LogicalPlan] = {
+      children: Seq[LogicalPlan],
+      dedupDistinct: Boolean): Seq[LogicalPlan] = {
```

Comment:
Ok, so I took a longer look. The current approach uses tail recursion with a bunch of state. I tried rewriting this as a tail recursive function without that much state, but that proved to be very painful. So it seems simpler to just use a purely imperative function:
```scala
private def flattenUnion(union: Union, flattenDistinct: Boolean): Union = {
    val stack = mutable.Stack[LogicalPlan](union)
    val flattened = mutable.ArrayBuffer.empty[LogicalPlan]
    while (stack.nonEmpty) {
      stack.pop() match {
        case Distinct(Union(children)) if flattenDistinct =>
          stack.pushAll(children)
        case Union(children) =>
          stack.pushAll(children)
        case child =>
          flattened += child
      }
    }
    Union(flattened)
  }
```
@gatorsmile why did you use `reverseMap`?